Static Par Timing Analysis

calebmkim opened the pull request:
Implements the idea in #1353. 

The key HashMap has the following structure: 
```
Maps Par ID -> (Map that maps Enable ID -> clock interval (i,j))
```
The intervals are relative to the start of the par block. 

Couple of points: 

1) I chose not to include invokes in this because they're not included in #1331. Although, if we wanted to, it shouldn't be too hard to add support for them in this analysis. 
2) If we have nested static pars (that merge-static-par pass doesn't deal with), like: 
```
@static par{
    @static seq{
            ...
            @static par{
                @static A; 
                 ...
             }
            ...     
    }
}
```
Then the enable A will appear in the hashmap relative to _both_  par blocks.

rachitnigam commented on calyx/src/analysis/static_par_timing.rs:
```diff
@@ -0,0 +1,192 @@
+
+/// maps the node ids of enables to a set of tuples (i,j), which is clock cycles (relative
+/// to the start of the par) that enable is live
+type ParTimingMap = HashMap<u64, HashSet<(u64, u64)>>;
```
There might be a smarter way to represent the inner HashSet; for example, if we sort the ranges, we can perform a binary search the list to figure out when something is active.

In general, maybe can think about a nicer interface for ParTimingMap to the outside world?

calebmkim commented:
I rewrote the data structure, although it's now quite specific to the cell sharing/ live range analysis stuff. The main difference is that it now maps threads (i.e., children) of par blocks to a map that maps cell names to a vector of intervals for which that cell is live. Doing this makes implementation in the actual cell sharing pass quite easy, so I added that.  

One option I was thinking is, as we need different data structures, just update the analysis struct to include more fields?

rachitnigam commented:
Sounds great! I'll try to review this later today. It occurs to me that another place we can test the utility of this analysis is in the MergeStaticPar pass which attempts to compute when things are scheduled and rearrange. In general, it could enable use to build a generalization of that pass that computes when things run and generates a control program that can be lowered much more efficiently. If you end up reading that pass and have thoughts on how to integrate this analysis with it, feel free to open an issue

rachitnigam commented on calyx/src/passes/cell_share.rs:
```diff
@@ -215,6 +234,9 @@ impl CellShare {
             .collect();
 
+        let print_par_timing =
+            given_opts.iter().any(|arg| *arg == "print_par_timing");
```
Why do you need to explicitly parse this? The `get_opts` function should be able to parse this for you. For example: https://github.com/cucapra/calyx/blob/master/calyx/src/passes/top_down_static_timing/tdst.rs#L751

calebmkim replied:
Ah, I'm just realizing, if I use `get_opts()` then it will give me a warning for flags like "-x cell-share:bounds=2,4,-1", since `get_opts()`.

rachitnigam commented on calyx/src/analysis/static_par_timing.rs:
```diff
@@ -0,0 +1,321 @@
+    /// if `par_id` is not static, then will automtically return true
+    pub fn liveness_overlaps(
+        &self,
+        par_id: &u64,
```
One common pattern used when building such interfaces is the [NewType Idiom](https://doc.rust-lang.org/rust-by-example/generics/new_types.html). Essentially, if your interface produces these `u64` tokens and consumes them, you can construct a new struct that wraps the `u64` and doesn't allow the user code to look into it. That way, the *only* thing user code can do is get these node and thread ID tokens from the analysis and send them back to perform various queries. 

The `ir/id.rs` shows how we use this for representing identifier using simple wrapped numbers.

calebmkim replied:
I think I understand what you're saying: we should have a struct `Par(u64)` and `ParThread(u64)`, and the `cell_share` pass should never actually look at the u64 that's inside these structs. 

One thing I should probably mention: in the cell-share pass, the `Par(u64)` and `ParThread(u64)` tokens would actually be generated from the LiveRangeAnalysis Analysis, whereas the `Par(u64)` and `ParThread(u64)` tokens we deal with here are obviously just generated by the ParStaticTiming Analysis. 
Currently,  I'm relying on the fact that `ControlId` generates a deterministic labeling of node IDs, so the labeling of IDs will be the same for both analyses. Do you think it would be more "solid" to change `ControlId` so that we can actually instantiate it, and so we could literally pass the same `ControlId` instance through both the Live Range Analysis and StaticParTiming Analysis? This wouldn't really affect correctness, but it does seem more "solid".

rachitnigam replied:
Yeah, I think having `ControlId` actually provide those "wrapped tokens" is definitely a way better interface. An even funkier trick that people use in rust is to actually associate all produced tokens with the analysis by giving it a fake lifetime:
```
pub struct WrappedIdx<'a> {
  idx: u64,
  _d: std::marker::PhantomData<'a>
}

pub struct LiveRanges<'a> {
  pub fn get_label(node: &ir::Id) -> WrappedIdx<'a> ...
}
```

rachitnigam commented on calyx/src/analysis/static_par_timing.rs:
```diff
@@ -0,0 +1,321 @@
+};
+use std::{collections::HashMap, fmt::Debug};
+
+use super::LiveRangeAnalysis;
```
Unnecessary space between imports

rachitnigam commented:
Minor comments in the PR review but a couple of major things we should consider doing after this is merged:
1. Redesigning the `ControlId` stuff so that we can provide a better interface. Currently, we have it implemented in a couple of different places with a couple of different interfaces, all of which use base `u64` to index into the analysis. Instead, we should come up with a way to build a [control-flow graph](https://en.wikipedia.org/wiki/Control-flow_graph) that tracks relationships between the groups and allows all of these passes to nicely query information.
2. As you mentioned, the queries supported by the analysis are currently very specific. It would be nice to generalize them. One way to do this is implementing a "static control rewriter" pass that combines the capabilities of `merge-static-par` and `static-conv-par` using queries over this analysis. A nice side effect will be that it solves #1340 as well.

If those two make sense, can you open issues talking about them? Apart from that, merge at your discretion!

rachitnigam commented:
Just noting that (1) is exactly what @paili0628 was attempting to do in https://github.com/cucapra/calyx/issues/1225

calebmkim commented:
Ok, merging. #1366 and #1367 flesh out what was discussed in this PR
